src/v/kafka/protocol/fetch.h

```diff
@@ -451,55 +452,67 @@ struct fetch_config {
     model::timeout_clock::time_point timeout;
     bool strict_max_bytes{false};
 };
-struct fetched_offset_range {
-    model::offset base_offset;
-    model::offset last_offset;
-};
+
 /**
- * Simple type aggregating either reader and offsets or an error
+ * Simple type aggregating either data or an error
  */
 struct read_result {
+    using foreign_data_t = ss::foreign_ptr<std::unique_ptr<iobuf>>;
+    using data_t = std::unique_ptr<iobuf>;
+    using variant_t = std::variant<data_t, foreign_data_t>;
     explicit read_result(error_code e)
       : error(e) {}
 
     read_result(
-      model::record_batch_reader rdr,
-      model::offset start_offset,
-      model::offset hw,
-      model::offset lso)
-      : reader(std::move(rdr))
-      , start_offset(start_offset)
-      , high_watermark(hw)
-      , last_stable_offset(lso)
-      , error(error_code::none) {}
-
-    read_result(
-      model::record_batch_reader rdr,
+      variant_t data,
       model::offset start_offset,
       model::offset hw,
       model::offset lso,
-      fetched_offset_range fr)
-      : reader(std::move(rdr))
+      std::vector<cluster::rm_stm::tx_range> aborted_transactions)
+      : data(std::move(data))
       , start_offset(start_offset)
       , high_watermark(hw)
       , last_stable_offset(lso)
-      , fetched_range(fr)
-      , error(error_code::none) {}
+      , error(error_code::none)
+      , aborted_transactions(std::move(aborted_transactions)) {}
 
     read_result(model::offset start_offset, model::offset hw, model::offset lso)
       : start_offset(start_offset)
       , high_watermark(hw)
       , last_stable_offset(lso)
       , error(error_code::none) {}
 
-    std::optional<model::record_batch_reader> reader;
+    bool has_data() const {
+        if (std::holds_alternative<data_t>(data)) {
+            return std::get<data_t>(data) != nullptr;
+        } else {
+            return !std::get<foreign_data_t>(data)->empty();
+        }
+    }
+
+    const iobuf& get_data() const {
+        if (std::holds_alternative<data_t>(data)) {
+            return *std::get<data_t>(data);
+        } else {
+            return *std::get<foreign_data_t>(data);
+        }
+    }
+
+    iobuf release_data() && {
+        if (std::holds_alternative<data_t>(data)) {
+            return std::move(*std::get<data_t>(data));
+        } else {
+            return std::move(*std::get<foreign_data_t>(data));
```

Comment:
it seems like this disables the benefit of foreign pointers, right? if the data is no longer owned by the foreign pointer then its destruction won't be handled by the foreign pointer (e.g. by sending a message back to the originating core).

Reply:
that could work. or create a reader that can read from foreign batch? that might be even faster if we know that the batch will be consume right away